src/controller/CHIPDeviceController.cpp

```diff
@@ -1588,43 +1588,59 @@ void DeviceCommissioner::SetupCluster(ClusterBase & base, DeviceProxy * proxy, E
-    report.Set<BasicVendor>(vendorId);
-    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
-}
+    CHIP_ERROR err;
+    CHIP_ERROR return_err = CHIP_NO_ERROR;
+    ReadCommissioningInfo info;
 
-void BasicProductCallback(void * context, uint16_t productId)
-{
-    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
-    CommissioningDelegate::CommissioningReport report;
-    report.Set<BasicProduct>(productId);
-    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
-}
+    // Using ForEachAttribute because this attribute can be queried on any endpoint.
+    err = mAttributeCache->ForEachAttribute(
+        app::Clusters::GeneralCommissioning::Id, [this, &info](const app::ConcreteAttributePath & path) {
+            if (path.mAttributeId != app::Clusters::GeneralCommissioning::Attributes::BasicCommissioningInfo::Id)
+            {
+                return CHIP_NO_ERROR;
+            }
+            TLV::TLVReader reader;
+            ReturnErrorOnFailure(this->mAttributeCache->Get(path, reader));
+            app::Clusters::GeneralCommissioning::Attributes::BasicCommissioningInfo::TypeInfo::DecodableType basicInfo;
+            ReturnErrorOnFailure(app::DataModel::Decode(reader, basicInfo));
+            info.general.recommendedFailsafe = basicInfo.failSafeExpiryLengthSeconds;
+            return CHIP_NO_ERROR;
+        });
 
-void BasicSoftwareCallback(void * context, uint32_t softwareVersion)
-{
-    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
-    CommissioningDelegate::CommissioningReport report;
-    report.Set<BasicSoftware>(softwareVersion);
-    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
-}
+    // Try to parse as much as we can here before returning, even if this is an error.
+    return_err = err == CHIP_NO_ERROR ? return_err : err;
 
-void AttributeReadFailure(void * context, CHIP_ERROR status)
-{
-    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
-    commissioner->CommissioningStageComplete(status);
-}
+    err = mAttributeCache->ForEachAttribute(app::Clusters::Basic::Id, [this, &info](const app::ConcreteAttributePath & path) {
+        TLV::TLVReader reader;
+        if (path.mAttributeId != app::Clusters::Basic::Attributes::VendorID::Id &&
+            path.mAttributeId != app::Clusters::Basic::Attributes::ProductID::Id &&
+            path.mAttributeId != app::Clusters::Basic::Attributes::SoftwareVersion::Id)
+        {
+            // Continue on
+            return CHIP_NO_ERROR;
+        }
 
-// AttributeCache::Callback impl
-void DeviceCommissioner::OnDone()
-{
-    NetworkClusters clusters;
+        ReturnErrorOnFailure(this->mAttributeCache->Get(path, reader));
+        switch (path.mAttributeId)
+        {
+        case app::Clusters::Basic::Attributes::VendorID::Id:
+            return app::DataModel::Decode(reader, info.basic.vendorId);
```

Comment:
Again, we can just Get() with the right types on the cache and no get the TLVReader separately....

The only reason we had to get TLVReader for FeatureMap is because we wanted it as a type that did not match the attribute type.
```suggestion
            return this->mAttributeCache->Get(path, info.basic.vendorId);
```
and similar for the others.

Reply:
Will change. Recommend adding more documentation to this interface - I was going off example code provided and there is no indication of which interface is preferred or why.